File: MyRecyclerView/app/src/main/java/com/solie/recyclerview/MainActivity.java
package com.solie.recyclerview;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Toast;

public class MainActivity extends AppCompatActivity {

    RecyclerView recyclerView;

    SingerAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = findViewById(R.id.recyclerView);

        LinearLayoutManager layoutManager = new LinearLayoutManager(this, LinearLayoutManager.HORIZONTAL, false);
        recyclerView.setLayoutManager(layoutManager);

        adapter = new SingerAdapter(getApplicationContext());

        adapter.addItem(new SingerItem("걸그룹1", "000-0000-0000"));
        adapter.addItem(new SingerItem("걸그룹2", "111-1111-1111"));
        adapter.addItem(new SingerItem("걸그룹3", "222-2222-2222"));

        recyclerView.setAdapter(adapter);

        adapter.setOnItemClickListener(new SingerAdapter.OnItemClickListener() {
            @Override
            public void onItemClick(SingerAdapter.ViewHolder holder, View view, int position) {
                SingerItem item = adapter.getItem(position);
                Toast.makeText(getApplicationContext(), "item 선택됨" + item.getName(), Toast.LENGTH_SHORT).show();
            }
        });

    }
}

File: MVVMPractice/app/src/main/java/com/solie/mvvmpractice/MainActivity.kt
package com.solie.mvvmpractice

import android.os.Bundle
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProviders
import androidx.recyclerview.widget.LinearLayoutManager
import com.solie.mvvmpractice.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    private lateinit var binding : ActivityMainBinding
    private var list = mutableListOf<Exam1>()
    lateinit var model : MyViewModel

    private val observerListener = Observer<Exam1> {
        it?.let { exam1 ->  }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        model = ViewModelProviders.of(this).get(MyViewModel::class.java)
        if(::model.isInitialized) {
            model.mExam1.observe(this, observerListener)
        }
        setRecycler()
    }

    private fun setRecycler() {
        val adapter = RecyclerAdapter(this)

        list.add(Exam1("가", 1))
        list.add(Exam1("나",2))
        list.add(Exam1("다",1))
        list.add(Exam1("라",2))

        binding.mainRecycler.layoutManager = LinearLayoutManager(this)
        binding.mainRecycler.adapter = adapter
        adapter.setData(list)
        adapter.cListener(object : RecyclerAdapter.CListener{
            override fun onClick(view: View, position: Int) {
                model.request(position)
            }

        })
    }
}